modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java

```diff
@@ -0,0 +1,340 @@
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.vectormosaic;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.geotools.data.DataStore;
+import org.geotools.data.DataUtilities;
+import org.geotools.data.FeatureReader;
+import org.geotools.data.FeatureSource;
+import org.geotools.data.Query;
+import org.geotools.data.simple.SimpleFeatureIterator;
+import org.geotools.data.simple.SimpleFeatureReader;
+import org.geotools.data.simple.SimpleFeatureSource;
+import org.geotools.data.store.ContentEntry;
+import org.geotools.data.store.ContentFeatureSource;
+import org.geotools.feature.SchemaException;
+import org.geotools.feature.simple.SimpleFeatureBuilder;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.filter.FilterAttributeExtractor;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.util.logging.Logging;
+import org.locationtech.jts.geom.Envelope;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.feature.type.AttributeDescriptor;
+import org.opengis.filter.Filter;
+
+/** FeatureSource for a vector mosaic. */
+public class VectorMosaicFeatureSource extends ContentFeatureSource {
+    static final Logger LOGGER = Logging.getLogger(VectorMosaicFeatureSource.class);
+    VectorMosaicTypeConfiguration config;
+
+    /**
+     * VectorMosaicFeatureSource constructor.
+     *
+     * @param contentEntry The content entry for the feature source.
+     * @param store The data store for the feature source.
+     * @param config The configuration for the feature source.
+     */
+    public VectorMosaicFeatureSource(
+            ContentEntry contentEntry,
+            VectorMosaicStore store,
+            VectorMosaicTypeConfiguration config) {
+        super(contentEntry, null);
+        this.config = config;
+    }
+
+    /**
+     * Get the FeatureStore for the feature source.
+     *
+     * @return The FeatureStore for the feature source.
+     */
+    public VectorMosaicStore getStore() {
+        return (VectorMosaicStore) getEntry().getDataStore();
+    }
+
+    @Override
+    protected ReferencedEnvelope getBoundsInternal(Query query) throws IOException {
+        List<ReferencedEnvelope> allBounds = new ArrayList<>();
+        for (VectorMosaicGranule granule : config.getGranules()) {
```

Comment:
Code in this class is mostly broken because of this pattern: it should split the filter based on the attributes that are available in the index, query the index only hitting the granules that are matching, and only after go and open the potential candidate granules...

Reply:
Fixed